relay-pattern/src/wildmatch.rs

```diff
@@ -0,0 +1,605 @@
+            .find(|&(_, c)| Self::ranges_match(c, negated, ranges))
+    }
+}
+
+/// A case sensitive [`Matcher`].
+struct CaseSensitive;
+
+impl Matcher for CaseSensitive {
+    #[inline(always)]
+    fn is_prefix(haystack: &str, needle: &Literal) -> Option<usize> {
+        let needle = needle.as_case_converted_bytes();
+        memchr::arch::all::is_prefix(haystack.as_bytes(), needle).then_some(needle.len())
+    }
+
+    #[inline(always)]
+    fn find(haystack: &str, needle: &Literal) -> Option<(usize, usize)> {
+        let needle = needle.as_case_converted_bytes();
+        memchr::memmem::find(haystack.as_bytes(), needle).map(|offset| (offset, needle.len()))
+    }
+
+    #[inline(always)]
+    fn ranges_match(c: char, negated: bool, ranges: &Ranges) -> bool {
+        ranges.contains(c) ^ negated
+    }
+}
+
+/// A case insensitive [`Matcher`].
+struct CaseInsensitive;
+
+impl Matcher for CaseInsensitive {
+    #[inline(always)]
+    fn is_prefix(haystack: &str, needle: &Literal) -> Option<usize> {
+        // We can safely assume `needle` is already full lowercase. This transformation is done on
+        // token creation based on the options.
+        //
+        // The haystack cannot be converted to full lowercase to not break class matches on
+        // uppercase unicode characters which would produce multiple lowercase characters.
+        //
+        // TODO: benchmark if allocation free is better/faster.
+        let needle = needle.as_case_converted_bytes();
+        let lower_haystack = haystack.to_lowercase();
+
+        memchr::arch::all::is_prefix(lower_haystack.as_bytes(), needle)
+            .then(|| recover_offset_len(haystack, 0, needle.len()).1)
+    }
+
+    #[inline(always)]
+    fn find(haystack: &str, needle: &Literal) -> Option<(usize, usize)> {
+        // TODO: implement manual lowercase which remembers if there were 'special' unicode
+        // conversion involved, if not, there is no recovery necessary.
+        // TODO: benchmark if a lut from offset -> original offset makes sense.
+        // TODO: benchmark allocation free and search with proper case insensitive search.
+        let needle = needle.as_case_converted_bytes();
+        let lower_haystack = haystack.to_lowercase();
+
+        let offset = memchr::memmem::find(lower_haystack.as_bytes(), needle)?;
+
+        // `offset` now points into the lowercase converted string, but this may not match the
+        // offset in the original string. Time to recover the index.
+        Some(recover_offset_len(haystack, offset, offset + needle.len()))
+    }
+
+    #[inline(always)]
+    fn ranges_match(c: char, negated: bool, ranges: &Ranges) -> bool {
+        let matches = exactly_one(c.to_lowercase()).is_some_and(|c| ranges.contains(c))
+            || exactly_one(c.to_uppercase()).is_some_and(|c| ranges.contains(c));
+        matches ^ negated
+    }
+}
+
+#[inline(always)]
+fn is_match_impl<T, M>(tokens: &T, haystack: &str) -> bool
```

Comment:
nit: Should we declare `is_match_impl` before `Matcher`? Then it nicely reads from top to bottom `is_match -> is_match_impl -> M::*`